resolver/upstream_tree_resolver_test.go

```diff
@@ -0,0 +1,281 @@
+			sutConfig = config.UpstreamsConfig{}
+		})
+
+		It("should return error", func() {
+			Expect(err).To(HaveOccurred())
+			Expect(err).To(MatchError(ContainSubstring("no external DNS resolvers configured")))
+			Expect(sut).To(BeNil())
+		})
+	})
+
+	When("has only default group", func() {
+		BeforeEach(func() {
+			sutConfig = config.UpstreamsConfig{
+				Groups: config.UpstreamGroups{
+					upstreamDefaultCfgName: {
+						{Host: "wrong"},
+						{Host: "127.0.0.1"},
+					},
+				},
+			}
+			branches = createBranchesMock(sutConfig)
+		})
+		Describe("Type", func() {
+			It("does not return error", func() {
+				Expect(err).ToNot(HaveOccurred())
+			})
+			It("follows conventions", func() {
+				expectValidResolverType(sut)
+			})
+			It("returns mock", func() {
+				Expect(sut.Type()).To(Equal("mock"))
+			})
+		})
+	})
+
+	When("has multiple groups", func() {
+		BeforeEach(func() {
+			sutConfig = config.UpstreamsConfig{
+				Groups: config.UpstreamGroups{
+					upstreamDefaultCfgName: {
+						{Host: "wrong"},
+						{Host: "127.0.0.1"},
+					},
+					"test": {
+						{Host: "some-resolver"},
+					},
+				},
+			}
+			branches = createBranchesMock(sutConfig)
+		})
+		Describe("Type", func() {
+			It("does not return error", func() {
+				Expect(err).ToNot(HaveOccurred())
+			})
+			It("follows conventions", func() {
+				expectValidResolverType(sut)
+			})
+			It("returns upstream_tree", func() {
+				Expect(sut.Type()).To(Equal(upstreamTreeResolverType))
+			})
+		})
+		Describe("Configuration output", func() {
+			It("should return configuration", func() {
+				Expect(sut.IsEnabled()).Should(BeTrue())
+
+				logger, hook := log.NewMockEntry()
+				sut.LogConfig(logger)
+				Expect(hook.Calls).ToNot(BeEmpty())
+			})
+		})
+
+		Describe("Name", func() {
+			var utrSut *UpstreamTreeResolver
+			JustBeforeEach(func() {
+				utrSut = sut.(*UpstreamTreeResolver)
+			})
+
+			It("should contain correct resolver", func() {
+				name := utrSut.Name()
+				Expect(name).ShouldNot(BeEmpty())
+				Expect(name).Should(ContainSubstring(upstreamTreeResolverType))
+			})
+		})
+
+		When("client specific resolvers are defined", func() {
+			BeforeEach(func() {
+				defaultMock := &mockResolver{}
+				defaultMock.On("Resolve", mock.Anything)
+				defaultMock.ResponseFn = func(req *dns.Msg) *dns.Msg {
+					res := new(dns.Msg)
+					res.SetReply(req)
+
+					ptr := new(dns.PTR)
+					ptr.Ptr = "default"
+					ptr.Hdr = util.CreateHeader(req.Question[0], 1)
+					res.Answer = append(res.Answer, ptr)
+
+					return res
+				}
+
+				clientSpecificExactMock := &mockResolver{}
+				clientSpecificExactMock.On("Resolve", mock.Anything)
+
+				clientSpecificWildcardMock := &mockResolver{}
+				clientSpecificWildcardMock.On("Resolve", mock.Anything)
+
+				clientSpecificIPMock := &mockResolver{}
+				clientSpecificIPMock.On("Resolve", mock.Anything)
+
+				clientSpecificCIRDMock := &mockResolver{}
+				clientSpecificCIRDMock.On("Resolve", mock.Anything)
```

Comment:
Tried to understand this test... It looks like these mock resolvers are not used, right?

Reply:
You're correct. I forgot to remove them after adding the `createMockResolver` func :sweat_smile: